### PR TITLE
 Add LocationResolver framework

### PR DESCRIPTION
**Pull request by gab1one**
The LocationResolverService manages LocationResolver plugins that
provide translation from URI to Location.

### CODE REVIEW

**Comment by gab1one** [comment]
@ctrueden I updated the code based on our discussion on Friday, should be good to merge now.

**Comment by gab1one** [comment]
@ctrueden I moved the implementation of the resolve method into the interface, as discussed.

File: src/main/java/org/scijava/io/location/resolve/AbstractLocationResolver.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ * #L%
+ */
+
+package org.scijava.io.location.resolve;
```
Comment (ctrueden) [inline src/main/java/org/scijava/io/location/resolve/AbstractLocationResolver.java]: Let's put this in `org.scijava.io.location` without the `resolve` subpackage. The service, while focused on location resolution right now, might do other things as well in the future; it is called `LocationService` after all and not `LocationResolverService`.

File: src/main/java/org/scijava/io/location/resolve/DefaultLocationService.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ * #L%
+ */
+
+package org.scijava.io.location.resolve;
```
Comment (ctrueden) [inline src/main/java/org/scijava/io/location/resolve/DefaultLocationService.java]: Move up one package level.

File: src/main/java/org/scijava/io/location/resolve/FileLocationResolver.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ * #L%
+ */
+
+package org.scijava.io.location.resolve;
```
Comment (ctrueden) [inline src/main/java/org/scijava/io/location/resolve/FileLocationResolver.java]: Move up one package level.

File: src/main/java/org/scijava/io/location/resolve/LocationResolver.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ * #L%
+ */
+
+package org.scijava.io.location.resolve;
```
Comment (ctrueden) [inline src/main/java/org/scijava/io/location/resolve/LocationResolver.java]: Move up one package level.

File: src/main/java/org/scijava/io/location/resolve/LocationService.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ * #L%
+ */
+
+package org.scijava.io.location.resolve;
```
Comment (ctrueden) [inline src/main/java/org/scijava/io/location/resolve/LocationService.java]: Move up one package level.

File: src/test/java/org/scijava/io/location/resolve/FileLocationResolverTest.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ * #L%
+ */
+
+package org.scijava.io.location.resolve;
```
Comment (ctrueden) [inline src/test/java/org/scijava/io/location/resolve/FileLocationResolverTest.java]: Move up one package level.

File: src/test/java/org/scijava/io/location/resolve/FileLocationResolverTest.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+		Location loc = resolver.resolve(uri);
+		assertTrue(loc instanceof FileLocation);
+		assertEquals(uri, loc.getURI().toString());
+	}
+
+	@Test
+	public void testURIResolve() throws URISyntaxException {
+		URI uri = new URI(getClass().getResource("/").toString());
+		Location loc = resolver.resolve(uri);
+		assertTrue(loc instanceof FileLocation);
```
Comment (ctrueden) [inline src/test/java/org/scijava/io/location/resolve/FileLocationResolverTest.java]: Does that really work always? It seems dicey. Maybe use `ClassUtils.getLocation` instead? Or just hardcode something like `new File(".").getAbsolutePath()` which is guaranteed to work on every platform?

File: src/test/java/org/scijava/io/location/resolve/LocationServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+	@Test
+	public void testResolve() throws URISyntaxException {
+		final Context ctx = new Context(LocationService.class);
+		final LocationService loc = ctx.getService(LocationService.class);
+
+		final URI uri = getClass().getResource("/").toURI();
+		final LocationResolver res = loc.getResolver(uri);
+
+		assertTrue(res instanceof FileLocationResolver);
```
Comment (ctrueden) [inline src/test/java/org/scijava/io/location/resolve/LocationServiceTest.java]: Again, not convinced that `getClass().getResource(...)` URIs will always end up as files. Let's be safer about this.

**Comment by gab1one** [comment]
@ctrueden I included your suggestions

**Pull request merged**